modules/example-browser/src/caching_cmm.ts

```diff
@@ -0,0 +1,215 @@
+   */
+  const maxBytesEncrypted = 100
+
+  /* The maximum number of messages that will be encrypted under a single data key.
+   * This value is optional,
+   * but you should configure the lowest value possible.
+   */
+  const maxMessagesEncrypted = 10
+
+  const cachingCMM = new WebCryptoCachingMaterialsManager({
+    backingMaterials: keyring,
+    cache,
+    partition,
+    maxAge,
+    maxBytesEncrypted,
+    maxMessagesEncrypted
+  })
+
+  /* Encryption context is a *very* powerful tool for controlling
+   * and managing access.
+   * When you pass an encryption context to the encrypt function,
+   * the encryption context is cryptographically bound to the data.
+   * If you don't pass in the same encryption context when decrypting,
+   * the decrypt function fails.
+   * The encryption context is ***not*** secret!
+   * Encrypted data is opaque.
+   * You can use an encryption context to assert things about the encrypted data.
+   * The encryption context helps you to determine
+   * whether the ciphertext you retrieved is the ciphertext you expect to decrypt.
+   * For example, if you are are only expecting data from 'us-west-2',
+   * the appearance of a different AWS Region in the encryption context can indicate malicious interference.
+   * See: https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#encryption-context
+   *
+   * Cached data keys are reused ***only*** when the encryption contexts passed into the functions are an exact case-sensitive match.
+   * See: https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/data-caching-details.html#caching-encryption-context
+   */
+  const encryptionContext = {
+    stage: 'demo',
+    purpose: 'simple demonstration app',
+    origin: 'us-west-2'
+  }
+
+  /* Find data to encrypt. */
+  const plainText = new Uint8Array([1, 2, 3, 4, 5])
+
+  /* Encrypt the data.
+   * The caching CMM only reuses data keys
+   * when it know the length (or an estimate) of the plaintext.
+   * However, in the browser,
+   * you must provide all of the plaintext to the encrypt function.
+   * Therefore, the encrypt function in the browser knows the length of the plaintext
+   * and does not accept a plaintextLength option.
+   */
+  const { result } = await encrypt(cachingCMM, plainText, { encryptionContext })
+
+  /* Log the plain text
+   * only for testing and to show that it works.
+   */
+  console.log('plainText:', plainText)
+  document.write('</br>plainText:' + plainText + '</br>')
+
+  /* Log the base64-encoded result
+   * so that you can try decrypting it with another AWS Encryption SDK implementation.
+   */
+  const resultBase64 = toBase64(result)
+  console.log(resultBase64)
+  document.write(resultBase64)
+
+  /* Decrypt the data.
+   * NOTE: This decrypt request will not use the data key
+   * that was cached during the encrypt operation.
+   * Data keys for encrypt and decrypt operations are cached separately.
```

Comment:
Just a question. Are the data keys cached separately or just not reused when the operations differ. 

Should I add "operation (encrypt, decrypt)" to this section that explains when cached data keys are reused?
https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/data-caching-details.html#how-caching-works

Reply:
The are cached under separate ID's.  So they are in the same cache, but they are intentionally not shared.